make/autoconf/lib-vmath.m4

```diff
@@ -0,0 +1,125 @@
+# accompanied this code).
+#
+# You should have received a copy of the GNU General Public License version
+# 2 along with this work; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+# or visit www.oracle.com if you need additional information or have any
+# questions.
+#
+
+###############################################################################
+#
+# Setup vmath framework and check its dependences
+#
+AC_DEFUN_ONCE([LIB_SETUP_VMATH],
+[
+  AC_ARG_WITH(libsleef, [AS_HELP_STRING([--with-libsleef],
+      [specify prefix directory for the libsleef library
+      (expecting the libraries under PATH/lib and the headers under PATH/include)])])
+
+  ENABLE_LIBSLEEF=false
+  LIBVMATH_CFLAGS=
+  LIBVMATH_LIBS=
+
+  if test "x$NEEDS_LIB_SLEEF" = xfalse; then
+    if test "x${with_libsleef}" != "x" &&
+       test "x${with_libsleef}" != "xno"; then
+      AC_MSG_WARN([[libsleef is not used, so --with-libsleef is ignored]])
+    fi
+  else
+    LIBSLEEF_FOUND=no
+
+    if test "x${with_libsleef}" = "xno"; then
+      AC_MSG_NOTICE([libvmath will not be built, because its dependence libsleef is disabled])
+    else
+      if test "x${with_libsleef}" != "x" &&
+         test "x${with_libsleef}" != "xyes"; then
+        # Check the specified libsleef
+        AC_MSG_CHECKING([for the specified LIBSLEEF])
+        if test -e ${with_libsleef}/lib/libsleef.so &&
+           test -e ${with_libsleef}/include/sleef.h; then
+          LIBSLEEF_FOUND=yes
+          LIBVMATH_LIBS="-L${with_libsleef}/lib"
+          LIBVMATH_CFLAGS="-I${with_libsleef}/include"
+          AC_MSG_RESULT([yes])
+        else
+          AC_MSG_RESULT([no])
+          AC_MSG_ERROR([Could not locate libsleef.so or sleef.h in ${with_libsleef}])
+        fi
+      else
+        # Check for the libsleef under system locations
+        # Do not try pkg-config if we have a sysroot set.
+        if test "x$SYSROOT" = "x" &&
+           test "x${LIBSLEEF_FOUND}" = "xno"; then
+          PKG_CHECK_MODULES([LIBSLEEF], [sleef], [LIBSLEEF_FOUND=yes], [LIBSLEEF_FOUND=no])
+        fi
+        if test "x${LIBSLEEF_FOUND}" = "xno"; then
+          AC_CHECK_HEADERS([sleef.h],
+              [LIBSLEEF_FOUND=yes],
+              []
+          )
+        fi
+
+        # Print error if user runs just with "--with-libsleef", but libsleef is not installed
+        if test "x${with_libsleef}" = "xyes" &&
+           test "x${LIBSLEEF_FOUND}" = "xno"; then
+          HELP_MSG_MISSING_DEPENDENCY([sleef])
+          AC_MSG_ERROR([Could not find libsleef! $HELP_MSG])
+        fi
+      fi
+
+      if test "x${LIBSLEEF_FOUND}" = "xyes"; then
+        ENABLE_LIBSLEEF=true
+        LIBVMATH_LIBS="${LIBVMATH_LIBS} -lsleef"
+
+        if test "x${OPENJDK_TARGET_CPU}" = "xaarch64"; then
+          # Check the ARM SVE feature
+          SVE_CFLAGS="-march=armv8-a+sve"
+
```

Comment:
What's this about? We're building a standard binary, and all SVE detection is to be deferred to runtime.

Reply:
We have to use this c-compiler option to build out the SVE ABIs (e.g. `svfloat32_t sinfx_u10sve(svfloat32_t input)`) in `libvmath.so`. Without this option, at build time, all the sve related featues like `arm_sve.h / __ARM_FEATURE_SVE` are missing, together with the sve symbols in `libvmath.so` we needed at runtime.  That means at runtime, hotspot cannot find out the sve symbols and the vector operations will fall back to the default java implementation.

Reply:
You still need to separate out the SVE detection from the libsleef code, and provide a way to enable/disable it from the configure command line. It is not okay to auto-detect if features should be turned on or off by default, but it should always be possible to override.

Reply:
Apparently the situation is this: If your build machine happens to have SVE, then you will get SVE support in the vmath library. The SVE support will be used during runtime if the machine you run on has SVE support.

If your build host happens to to not have SVE, then the vmath library will be built without SVE support, and no matter if your runtime machine has SVE or not, it will not provide SVE support in the vmath library.

Now, if your CI farm has an arbitrarily selection of aarch64 machines with and without SVE, then you have no idea what you are going to get in your build.

We have been very careful in staying clear of this kind of "random" build system behavior. The system you build on should not affect the output -- at least, not without a chance to override the default value.

In fact, I am not even sure why it seems to the PR author to be a good idea to let the default be dependent on the build machine at all. My personal opinion is that it would be better to select either "SVE enabled" or "SVE disabled" as the default, and then let the user override this on the configure command line, if they target a platform with different SVE availability.

Reply:
> That's fine, but we must make sure that SVE is not used by the compiler in any other places. If you've already built on non-SVE and tested the result on both SVE and non-SVE, I'm happy.

Agree. 

Since it contains both NEON and SVE functions in libvmath.so, we have to disable SVE feature when building those NEON functions. We want to separate NEON/SVE functions in two files, build them with different cflags (i.e. only build SVE sources with `-march=armv8-a+sve`), and then link to the single `libvmath.so`. Do we have such similar examples or functions in current jdk make system? I'm still struggling on finding out an effective way for it.

> SVE support should be enabled regardless of the build machine. The same binary must run on both SVE and non-SVE machines, using SVE if it is advantageous. I suppose some ancient C++ compilers without SVE support still exist, but I see no very good reason to support them building JDK 22+.
> 
> Making a configure option to disable SVE support for vector math is a mistake, but IMO mostly harmless because no-one will ever turn it off.

Yes, SVE feature is also always enabled in jdk hotspot on SVE machines. If we add the option to give people disable SVE, it's weird that we disabling the SVE just in libvmath.so, and enabling it in hotspot. Besides, we choose the NEON stubs for smaller than 128-bit vector operations no matter whether the runtime machine supports SVE or not. So performance may not be an issue. Hence, I don't think people have reason disabling SVE features.

Reply:
It makes no sense to configure any of this at build time. Postpone all of the decisions to runtime.

Don't touch the make system.Instead, try to open the library at runtime with `os::dll_open()`, after (or inside) `VM_Version::initialize()`.

If you're not running on an SVE system, none of the SVE routines will be called, so it doesn't matter, right?